src/Microsoft.PowerShell.Security/security/CertificateProvider.cs

```diff
@@ -2633,42 +2725,132 @@ private bool IncludeArchivedCerts()
         }
 
-        // If it's Win8 or above, filter matching for certain properties is done by
-        // the certificate enumeration filter at the API level. In that case,
-        // filter.Purpose will be 'None' and MatchesFilter will return 'True'.
-        private static bool MatchesFilter(X509Certificate2 cert,
-                                   CertificateFilterInfo filter)
+        private static bool MatchesFilter(X509Certificate2 cert, CertificateFilterInfo filter)
         {
-            //
-            // no filter means, match everything
-            //
-            if ((filter == null) ||
-                (filter.Purpose == CertificatePurpose.NotSpecified) ||
-                (filter.Purpose == CertificatePurpose.All))
+            // No filter means, match everything
+            if (filter == null)
             {
                 return true;
             }
 
+            if (filter.Expiring > DateTime.MinValue && !SecuritySupport.CertIsExpiringInDays(cert, filter.Expiring))
+            {
+                return false;
+            }
+
+            if (filter.DnsName != null && !CertContainsName(cert, filter.DnsName))
+            {
+                return false;
+            }
+
+            if (filter.Eku != null && !CertContainsEku(cert, filter.Eku))
+            {
+                return false;
+            }
+
+            if (filter.SSLServerAuthentication && !CertIsSSLServerAuthentication(cert))
+            {
+                return false;
+            }
+
             switch (filter.Purpose)
             {
                 case CertificatePurpose.CodeSigning:
-                    if (SecuritySupport.CertIsGoodForSigning(cert))
-                    {
-                        return true;
-                    }
+                    return SecuritySupport.CertIsGoodForSigning(cert);
+
+                case CertificatePurpose.DocumentEncryption:
+                    return SecuritySupport.CertIsGoodForEncryption(cert);
 
+                case CertificatePurpose.NotSpecified:
+                case CertificatePurpose.All:
+                    return true;
+
+                default:
                     break;
+            }
 
-                case CertificatePurpose.DocumentEncryption:
-                    if (SecuritySupport.CertIsGoodForEncryption(cert))
+            return false;
+        }
+
+        /// <summary>
+        /// Check if the specified certificate has the name in DNS name list.
+        /// </summary>
+        /// <param name="cert">Certificate object.</param>
+        /// <param name="pattern">Wildcard pattern for DNS name to search.</param>
+        /// <returns>True on success, false otherwise.</returns>
+        internal static bool CertContainsName(X509Certificate2 cert, WildcardPattern pattern)
+        {
+            var list = (new DnsNameProperty(cert)).DnsNameList;
+            foreach (var dnsName in list)
+            {
+                if (pattern.IsMatch(dnsName.Unicode))
```

Comment:
Will this overmatch?

Reply:
We have a test for the filter and it works. Or I don't understand you question?

Reply:
You are using a regex?  Is there a possibility of overmatching.  such as `foo.bar.com` matching `foo1bar.com`

Reply:
I looked up the RFC and this should not overmatch

Reply:
Docs say that Wildcards * are permitted so it is possible.

Reply:
Do you ask about escaping?

Reply:
* wouldn't be over matching, as it's not allowed in the DNS Name.